README.md
# numbers.core/say

Clojure code to "read out" numbers between 1 and 1000 inclusive.

```Clojure
numbers.core/say
([n])
  returns the number `n` ∈ [1,1000] in English words
Spec
  args: (cat :n :numbers.core/nameable-num)
  ret: (and string? num-word?)
```

## Installation

Clone this repo, and use [Leiningen](https://leiningen.org/) to build the JAR files:

```
    $ lein uberjar
    $ chmod +x ./numbers.sh
```

## Usage

### Standalone

```
    $ ./numbers.sh 666
```

### Within clojure REPL

```Clojure
(require '[numbers.core :as n])

(n/say 666) ; => "six hundred and sixty-six"

```

## Design and Dev Notes

The `nums` map contains a (near) minimal set of names that are needed to construct all the numbers. These include the digits, and then the non-standard numbers above that. All other numbers can be constructed by adding "teen" or "ty" to a digit name (and then appending the ones if appropriate), and by counting the hundreds.

For invalid inputs outside the range, an appropriate string message is returned. This could be replaced by returning errors or by throwing an exception.

I decided that a single function `say` was better than a number of functions for 21-99 and 101-1000, which would require a `(declare say)` due to mutual recursion and would spread the simple logic over multiple functions. This also encouraged a minimal approach for each case in the main `cond` expression.

Unit testing this function *per se* was not really feasible, so only a regression test against previously computed values over key ranges is present. The regression test can be checked by changing appropriate entries in `nums`.

For completeness, a `spec` for the function's input and return values was included, with a `test.check` generative test. `::nameable-num` checks that the argument is between 1 and 1000 inclusive. `num-word?` checks that the return value contains at least one of the names in `nums`. This test can be checked for by changing types of a key in `nums`.

Development used Atom with [proto-repl](https://atom.io/packages/proto-repl), [humane-test-output](https://github.com/pjstadig/humane-test-output) and [lein-test-refresh](https://github.com/jakemcc/lein-test-refresh). Illustrative config for these is included in `project.clj` rather than in `~/.lein/profiles.clj`.

To rerun tests on every saved code change, run `lein test-refresh`. Note that [terminal-notifier](https://github.com/julienXX/terminal-notifier) is being used with the current `project.clj`.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-05-13
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-05-13
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/numbers/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/numbers/compare/0.1.0...0.1.1


test/numbers/core_test.clj
(ns numbers.core-test
  (:require [clojure.test :refer :all]
            [numbers.core :refer :all]
            [clojure.spec.test.alpha :as stest]
            [clojure.spec.gen.alpha :as gen]))

(def computed
  {[0 20]
   ["0 is not a natural number"
    "one" "two" "three" "four" "five"
    "six" "seven" "eight" "nine" "ten"
    "eleven" "twelve" "thirteen" "fourteen" "fifteen"
    "sixteen" "seventeen" "eighteen" "nineteen"]
   [50 72]
   ["fifty" "fifty-one" "fifty-two" "fifty-three" "fifty-four" "fifty-five"
    "fifty-six" "fifty-seven" "fifty-eight" "fifty-nine" "sixty"
    "sixty-one" "sixty-two" "sixty-three" "sixty-four" "sixty-five"
    "sixty-six" "sixty-seven" "sixty-eight" "sixty-nine" "seventy"
    "seventy-one"]
   [97 115]
   ["ninety-seven" "ninety-eight" "ninety-nine" "one hundred"
    "one hundred and one" "one hundred and two" "one hundred and three"
    "one hundred and four" "one hundred and five" "one hundred and six"
    "one hundred and seven" "one hundred and eight" "one hundred and nine"
    "one hundred and ten" "one hundred and eleven" "one hundred and twelve"
    "one hundred and thirteen" "one hundred and fourteen"]
   [998 1002]
   ["nine hundred and ninety-eight" "nine hundred and ninety-nine"
    "one thousand" "1001 is over 1000"]})

(deftest spec-test
  (testing "say conforms to its spec"
    (is (= nil (-> (stest/check `say) first :failure)))))

(deftest regression-test
  (testing "Names for key ranges of numbers"
    (are [x y] (= x y)
         (computed [0 20])     (mapv say (range 0 20))
         (computed [50 72])    (mapv say (range 50 72))
         (computed [97 115])   (mapv say (range 97 115))
         (computed [998 1002]) (mapv say (range 998 1002)))))


project.clj
(defproject numbers "0.1.0"
  :description "read numbers between 1 and 100 in English"
  :url "http://example.com/FIXME"
  :dependencies [[org.clojure/clojure "1.10.0"]]

  :main ^:skip-aot numbers.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}
             :dev {:dependencies [[proto-repl "0.3.1"]
                                  [pjstadig/humane-test-output "0.9.0"]
                                  [org.clojure/test.check "0.9.0"]]
                   :plugins [[com.jakemccrary/lein-test-refresh "0.24.1"]]
                   :injections [(require 'pjstadig.humane-test-output)
                                (pjstadig.humane-test-output/activate!)]
                   :test-refresh {:notify-command ["terminal-notifier" "-title" "Tests" "-message"]
                                     :quiet true
                                     :changes-only true}}})


src/numbers/core.clj
(ns numbers.core
  ^{:doc "Numbers in English"
    :author "Fergal Byrne"}
  (:require [clojure.spec.alpha :as s])
  (:gen-class))

(s/def ::nameable-num (s/and int? #(< 0 % 1001)))

(def ^:private nums {"one" 1 "two" 2 "three" 3 "four" 4 "five" 5
                     "six" 6 "seven" 7 "eight" 8 "nine" 9 "ten" 10
                     "eleven" 11 "twelve" 12 "thirteen" 13 "fifteen" 15
                     "eighteen" 18
                     "twenty" 20 "thirty" 30 "forty" 40 "fifty" 50
                     "eighty" 80 "one hundred" 100 "one thousand" 1000})

(def ^:private names (zipmap (vals nums) (keys nums)))

(defn num-word?
  "valid numbers contain at least one of the words in names"
  [n]
  (re-find (re-pattern (clojure.string/join "|" (keys nums))) n))

(s/fdef say
  :args (s/cat :n ::nameable-num)
  :ret  (s/and string? num-word?))

(defn say
  "returns the number `n` ∈ [1,1000] in English words"
  [n]
  (or (names n)
      (cond
        (< n 1)    (str n " is not a natural number")
        (< n 20)   (str (say (- n 10)) "teen")
        (< n 100)  (let [tens (quot n 10) ones (mod n 10)]
                     (str (or (names (* 10 tens))
                              (str (say tens) "ty"))
                          (when (pos? ones) (str "-" (say ones)))))
        (< n 1000) (str (say (quot n 100)) " hundred"
                        (when (pos? (mod n 100))
                          (str " and " (say (mod n 100)))))
        :else (str n " is over 1000"))))

(defn -main
  "Read out a number in English"
  [& args]
  (if (empty? args)
    (println "please provide a number between 1 and 1000")
    (try (let [n (Integer/parseInt (first args))]
           (println (say n)))
      (catch NumberFormatException e
        (println (str "Error: non-integer input '" (first args) "'"))))))
